### ISSUE TITLE
Behavior class handling of itemized-deduction response

### ISSUE THREAD

**Issue by martinholmer** [open-source-economics/Tax-Calculator#789]
After pull request #765 was merged into the master branch, the main issue with the Behavior class is the concern of @MattHJensen about how best to change itemized deductions by the itemized-deduction share of the total dollar change produced by the sum of the income and substitution effects. Right now the code is this:
```
calc.records.e19570 = np.where(ided > 0.,
                               calc.records.e19570 + delta_ided,
                               0.)
# TODO, we should create a behavioral modification variable instead of using e19570
```
The problem with adding `delta_ided` to the `e19570` variable is that `e19570` seems to not always be used in the tax-calculation code. For example, in the `ItemDed` function in the `functions.py` file, we have [these lines of code](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/functions.py#L309-L313) that by-pass the `e19570` variable when `puf` is True (which I'm assuming is the case when we are using the `puf.csv` file as Tax-Calculator input).

I'm not sure we should be adding a new "behavioral modification variable" as the TODO comment suggests. But we should think about which one of the existing `e` variables should have the `delta_ided` amount added to it.

Then in response @MattHJensen said:
> I am not sure about the right thing to do.
>
> At first I was thinking that we should add the feedback to a new variable that feeds into itemized deduction because a new variable would not be subject to potential policy reforms (any existing e variable is subject to reforms). But then I started second guessing whether that is actually the right objective. What about the case where someone eliminates all itemized deductions with haircuts -- in that case, we would want the new deductible expenses resulting from the feedback to be subject to those reforms, right? And if so, does that mean the best option would be to spread the feedback across all of the deductible expense e variables?

**Comment by feenberg** [comment]
On Tue, 14 Jun 2016, Martin Holmer wrote:

> 
> After pull request #765 was merged into the master branch, the main issue with the
> Behavior class is the concern of @MattHJensen about how best to change itemized
> deductions by the itemized-deduction share of the total dollar change produced by the
> sum of the income and substitution effects. Right now the code is this:
> 
> calc.records.e19570 = np.where(ided > 0.,
>                                calc.records.e19570 + delta_ided,
>                                0.)
> # TODO, we should create a behavioral modification variable instead of using e19570
> 
> The problem with adding delta_ided to the e19570 variable is that e19570 seems to not
> always be used in the tax-calculation code. For example, in the ItemDed function in
> the functions.py file, we have these lines of code that by-pass the e19570 variable
> when puf is True (which I'm assuming is the case when we are using the puf.csv file as
> Tax-Calculator input).
> 
> I'm not sure we should be adding a new "behavioral modification variable" as the TODO
> comment suggests. But we should think about which one of the existing e variables
> should have the delta_ided amount added to it.
> 
> Then in response @MattHJensen said:
>
>       I am not sure about the right thing to do.
>
>       At first I was thinking that we should add the feedback to a new variable
>       that feeds into itemized deduction because a new variable would not be
>       subject to potential policy reforms (any existing e variable is subject to
>       reforms). But then I started second guessing whether that is actually the
>       right objective. What about the case where someone eliminates all itemized
>       deductions with haircuts -- in that case, we would want the new deductible
>       expenses resulting from the feedback to be subject to those reforms,
>       right? And if so, does that mean the best option would be to spread the
>       feedback across all of the deductible expense e variables?
>

In the past I have always subtracted the behavioral change in itemized 
deductions from "other deductions" and allowed that to go negative. It is 
an arbitrary allocation, but no more so than any other. The difference in 
tax would be second-order, in any case. Spreading the change in deductions 
around is probably better if you have the time to do it, but it raises the 
question of deductions with floors. If you reduce medical below 10% of 
AGI, the effect on taxable income won't be what you need for the formula.

dan

**Comment by martinholmer** [comment]
Thinking more about this issue #789 has lead me to the conclusion that the current logic in the [Behavior `_update_ordinary_income` method](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/behavior.py#L179-L212) is incorrect.  Let me explain how I reached that conclusion.

Focus on those with itemized deductions (ID) because the logic is correct for those without itemized deductions.  Also, in what follows there is no division of the change in AGI into earnings and other income, the logic of which seems just fine.  I'm excluding that detail in order to simplify the exposition.

The logical relation between taxable income (TI) and AGI and ID is as follows:
```
(1) TI = AGI - ID - EX
```
where `EX` denotes personal exemptions.  The Behavior class leaves `EX` unchanged, so applying deltas to equation (1) produces equation (2):
```
(2) dTI = dAGI - dIT
```
Equation (2) is the logical relationship between these three change (or delta) amounts.

Now consider the logic in the `_update_ordinary_income` method.  If we let `a` denote the fraction `AGI / (AGI + ID)`, then we have from the code:
```
(3) dAGI = dTI * a
```
and 
```
(4) dID = dTI * (1-a)
```
So, if the code logic --- that is, equations (3) and (4) --- were correct, we should be able to substitute equations (3) and (4) into the right-hand side of equation (2) and simplify the resulting equation down to an identity (where the left-hand side, `dTI`, is equal to the right-hand side).  Doing the substitutions into equation (2) produces:
```
(5) dTI = dTI * a - dTI * (1-a)
```
Simplifying the right-hand side of equation (5) produces:
```
(6) dTI = dTI * a - dTI + dTI * a = 2 * a * dTI - dTI
```
So, when `a` is less than one (that is, there are itemized deductions), the equality in equation (6) is not satisfied.  In fact, the equality in equation (6) can be satisfied only if `a = 1` (that is, there are no itemized deductions).

**Comment by martinholmer** [comment]
Dan @feenberg said:
> In the past I have always subtracted the behavioral change in itemized deductions from "other deductions" and allowed that to go negative. It is an arbitrary allocation, but no more so than any other. The difference in tax would be second-order, in any case. Spreading the change in deductions
around is probably better if you have the time to do it, but it raises the question of deductions with floors. If you reduce medical below 10% of AGI, the effect on taxable income won't be what you need for the formula.

Thanks for your comment, Dan.  By "other deductions" do you mean the variable in the IRS-SOI PUF called E20400, which is described (in the 2009 IRS-SOI PUF) as "Miscellaneous deductions subject to AGI limitation, total"?

@MattHJensen @feenberg

**Comment by feenberg** [comment]
On Mon, 20 Jun 2016, Martin Holmer wrote:

> 
> Dan @feenberg said:
>
>       In the past I have always subtracted the behavioral change in itemized
>       deductions from "other deductions" and allowed that to go negative. It is
>       an arbitrary allocation, but no more so than any other. The difference in
>       tax would be second-order, in any case. Spreading the change in deductions
>       around is probably better if you have the time to do it, but it raises the
>       question of deductions with floors. If you reduce medical below 10% of
>       AGI, the effect on taxable income won't be what you need for the formula.
> 
> Thanks for your comment, Dan. By "other deductions" do you mean the variable in the
> IRS-SOI PUF called E20400, which is described (in the 2009 IRS-SOI PUF) as
> "Miscellaneous deductions subject to AGI limitation, total"?

I mean the one not subject to AGI limitation.

dan

**Comment by martinholmer** [comment]
Dan @feenberg said:
> I mean the one not subject to AGI limitation.

And what is the 2009 IRS-SOI PUF variable name for that one?

@MattHJensen @feenberg

**Comment by martinholmer** [comment]
Issue #789 has been resolved by the merger of pull request #796 into the master branch.
